lib/iris/fileformats/pp.py

```diff
@@ -1955,23 +1955,33 @@ def _convert_constraints(constraints):
     constraints = iris._constraints.list_of_constraints(constraints)
     pp_constraints = {}
     unhandled_constraints = False
+
+    def make_func(stashobj):
+        return lambda stash: stash==stashobj
+
     for con in constraints:
         if isinstance(con, iris.AttributeConstraint) and \
                 list(con._attributes.keys()) == ['STASH']:
             # Convert a STASH constraint.
+            # The attribute can be a STASH object, a stashcode string,or a 
+            # callable.
             stashobj = con._attributes['STASH']
-            if not isinstance(stashobj, STASH):
-                # The attribute can be a STASH object, or a stashcode string.
-                stashobj = STASH.from_msi(stashobj)
+            if callable(stashobj):
+                call_func = stashobj
+            else:
+                call_func = make_func(stashobj)
+
+            # Raise an exception if stashobj was not string or STASH object??   
```

Comment:
that sounds like the kind of thing someone would pass in, not realising it wouldn't work.  I'd raise an exception, with a nice message explaining to the user how to input 

something like
```
raise ValueError("Please input a string, of the form "mNNsNNiNNN" or STASH object")
```

for extra marks, you could add code to take an integer and convert it to a STASH object, but even then you might want an exception in the code here

All this said, you could make a case that your fixing a bug, and this behaviour was already here.

my view: raise an exception